Player and Dealer Implementation

nwarsalee opened the pull request:
Implemented the Player and the Dealer classes for the Blackjack simulator.

Both have a compute_play() function that is their main use.
This function will return an enum representing their decision based on their blackjack hand.

Both classes include additional functions and members to aid in the calculation of the decision.

 - [x] @esulu
 - [ ]  @cailynse

nwarsalee committed:
Create Dealer and Player files

Created Dealer and Player files.

Added initial helper functions to the Player class to help with reading and using their hand.

nwarsalee committed:
Implement Decision Making for Player and Dealer

Implemented the decision logic that is to be used by both the player and the dealer.

Player utilizes Basic Strategy rules.
Dealer utilizes dealer's default strategy of stand on 17 above.

Added more helper functions to Dealer, that are similar if not the same as the ones in Player.

Now both Player and Dealer make use of enum defined in definitions.py.

nwarsalee committed:
Add Print of Person hand and decision

Added a print function in both the Player and Dealer class that will print their given hand and the decision that was determined based on their respective strategies.

Added an attribute to both Dealer and Player called dev_mode.
Default value for this is set to False.
When set to True, it will activate all print and debug statements for the class (currently only activates the print_decision function)

nwarsalee committed:
Change Ace point addition from 10 to 11

Fixed problem where it was adding 10 onto the hand total instead of 11 for an Ace.

nwarsalee committed:
Remove is_pair function from Dealer

Removed is_pair function from the Dealer class.

Doesn't ever seem to be used.

nwarsalee committed:
Update hand_total comment

nwarsalee committed:
Rework Basic Strategy and Use PyDealer

Reworked basic strategy logic that player uses  to be in line with what is in the README for the project.

Updated both the Player and the Dealer class to make use of the PyDealer module for cards.

Both now expect a list of Card objects when computing their decision.

Helper functions have been modified to conform with PyDealer's API and created new helper functions where needed.

nwarsalee committed:
Fix Soft Hit case

Fixed error that had player standing when they should be hitting.

nwarsalee committed:
Add Blackjack enum and case for blackjack

Added a new blackjack enum for the Actions enum.

Added another check in the player and dealer's compute strategy to check for a blackjack.

nwarsalee committed:
Modify Player and Dealer

Added self argument to all Player and Dealer class functions.

Made some tweaks to functions to fix bugs.

nwarsalee committed:
Create test-actors file for testing

Created test-actors.py file for minor testing.

Updated gitignore to ignore pycache files.

nwarsalee committed:
Fix Bug in Dealer hand_total func

Fixed bug in hand_total function where it was passing in the wrong type of argument into value_to_int (was passing in the string of the value, not the actual card object itself).

nwarsalee committed:
Add tests compute_play

Added cases for all possible 2 card Blackjack hands and for all possible dealer upcards.

Tests the player's compute_play function with all these possible scenarios. Outputs the results to a file called test_player_basic_strategy.txt.

Test dealer's strategy with a 2 card blackjack hand. Outputs the results to a file called test_dealer_strategy.txt.

nwarsalee commented:
Here is a sample of the testing file that is created for the player.

> -----------[ Player's Hand: 2 of Hearts | 2 of Clubs  (Pairs) ]-----------
> 
> Player's Hand: 2 of Hearts | 2 of Clubs  (Pairs)
> Dealer's Upcard: 2 of Diamonds
> Decision: Actions.SPLIT
> 
> Player's Hand: 2 of Hearts | 2 of Clubs  (Pairs)
> Dealer's Upcard: 3 of Clubs
> Decision: Actions.SPLIT
> 
> Player's Hand: 2 of Hearts | 2 of Clubs  (Pairs)
> Dealer's Upcard: 4 of Spades
> Decision: Actions.SPLIT
> 
> Player's Hand: 2 of Hearts | 2 of Clubs  (Pairs)
> Dealer's Upcard: 5 of Clubs
> Decision: Actions.SPLIT
> 
> Player's Hand: 2 of Hearts | 2 of Clubs  (Pairs)
> Dealer's Upcard: 6 of Clubs
> Decision: Actions.SPLIT
> 
> Player's Hand: 2 of Hearts | 2 of Clubs  (Pairs)
> Dealer's Upcard: 7 of Spades
> Decision: Actions.SPLIT
> 
> Player's Hand: 2 of Hearts | 2 of Clubs  (Pairs)
> Dealer's Upcard: 8 of Clubs
> Decision: Actions.HIT
>

nwarsalee commented:
Here is a sample of the testing file created for the dealer.

> -----------[ Dealer's Hand: 2 of Hearts | 2 of Clubs  (Pairs) ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Clubs | 3 of Hearts ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Hearts | 4 of Diamonds ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Spades | 5 of Spades ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Spades | 6 of Spades ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Hearts | 7 of Diamonds ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Spades | 8 of Clubs ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Spades | 9 of Clubs ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Hearts | 10 of Clubs ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Clubs | Jack of Diamonds ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Hearts | Queen of Hearts ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Clubs | King of Spades ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 2 of Hearts | Ace of Hearts ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 3 of Diamonds | 2 of Spades ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 3 of Clubs | 3 of Hearts  (Pairs) ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 3 of Spades | 4 of Clubs ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 3 of Hearts | 5 of Clubs ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 3 of Clubs | 6 of Diamonds ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 3 of Hearts | 7 of Clubs ]-----------
>    Decision: Actions.HIT
> 
> -----------[ Dealer's Hand: 3 of Spades | 8 of Spades ]-----------
>    Decision: Actions.HIT
>

esulu commented on player.py:
```diff
@@ -0,0 +1,340 @@
+
+        # Case for Soft Hands
+        else:
+            if total in range(13, 17):
+                if dealer_up_val in range(4, 7):
+                    decision = definitions.Actions.DOUBLE
+                else:
+                    decision = definitions.Actions.HIT
+            
+            if total == 17:
+                if dealer_up_val in range(2, 7):
+                    decision = definitions.Actions.DOUBLE
+                else:
+                    decision = definitions.Actions.HIT
+
+            if total == 18:
+                if dealer_up_val in range(3, 7):
+                    decision = definitions.Actions.DOUBLE
+                elif dealer_up_val in range(9, 11):
+                    decision = definitions.Actions.HIT
+                else:
+                    decision = definitions.Actions.STAND
+            
+            if total == 19:
+                if dealer_up_val == 6:
+                    decision = definitions.Actions.DOUBLE
```
I think that the compute_play function in the player class will require a boolean value to keep track if there are split aces. This can be determined in the game class which can call the function with this value. 

It would be easier for the game class if the soft Dh or Ds options on the basic strategy table did not return double in the case of split aces.

nwarsalee committed:
Add split ace cases in Player compute

Added new optional argument for compute_play function in the Player class that takes in whether there were Aces that were split previously.

 Modified basic strategy logic to check for split aces in the case of Dh or Ds (Double if allowed, otherwise hit or split).

Changes should make logic in Game class easier for handling case where there are split aces.

nwarsalee committed:
Merge branch 'main' into player-and-dealer
